### PR TITLE
New Post: CSRF - protecting ifixit.com

### PR DESCRIPTION
**Pull request by danielbeardsley**
Describes the problem, typical prevention methods, and our simple method.

### CODE REVIEW

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+immediately before a form is submitted.
+The server then compares the form field with the cookie
+and ignores the request if they don't match.
+This frees us from the hassle of ensuring *every* form 
+includes a hidden csrf `<input>` in *every* template.
+
+## What does a CSRF attack allow?
+Cross-site request forgery allows an attacker
+to make arbitrary requests to a target site
+on behaf of another user,
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: behaf -> behalf

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+*with* their privleges but *without* their consent.
+
+## How is a CSRF attack executed?
+An attacker first figures out what is sent during a
+a particular POST request on the target site
+(maybe something important like deleting a Guide).
+Next, they create a similar form (via html or javascript) on their own site.
+When a user who is logged in on the target site visits the attackers site,
+it submits the form to the target site
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: `it` seems a little unclear here. I would simply change it to what exactly is submitting the form to the target site.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+At the very least, that's just more boiler-plate code
+that slows development progress.
+At the worst, you may forget to add the CSRF `<input>` tag
+to some seldom-used forms and they'll efectively be broken.
+
+### Our approach to prevention
+We use javascript to dynamically set a cookie
+and add an `<input>` element to the form
+immediately before it's submitted.
+Because we just have to prove that we can alter or read cookies
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: You need a comma after cookies.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+we don't need to depend on the server-side for setting the cookie,
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: I would change this comma to a semicolon.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+Here's the entire javascript implementation of our CSRF protection (Github [gist][gist]):
+{% gist 6060418 %}
+
+*Note: we use Mootools, but refactoring for no jQuery or no library would be
+fairly straightforward.*
+
+## How we deployed this
+We designed it, tested it extensively, ran our test suite many times, then
+soft-deployed it, silently reporting CSRF failures to a database. Once we'd
+worked out a few issues and were confident that we'd covered all our bases we
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: You need a comma after bases.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+flipped the switch on enforcement and it's been working well since.
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: You need a comma after enforcement.

**Comment by danielbeardsley** [comment]
@brotatos Thanks for the proofread! I've made all of those corrections.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+title: "CSRF - How we protect ifixit.com from request forgery"
+author: Daniel Beardsley
+author_url: http://github.com/danielbeardsley
+summary: We've come up with a minimalist method of protecting our site
+         from cross site request forgery.
+---
+
+We've come up with a minimalist method of
+protecting our site from cross site request forgery ([CSRF][csrf])
+and we'd like to share how we do it.
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: You need a comma before `and`.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+The server then compares the form field with the cookie
+and ignores the request if they don't match.
+This frees us from the hassle of ensuring *every* form 
+includes a hidden csrf `<input>` in *every* template.
+
+## What does a CSRF attack allow?
+Cross-site request forgery allows an attacker
+to make arbitrary requests to a target site
+on behalf of another user,
+*with* their privleges but *without* their consent.
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: privleges -> privileges

I don't think you need the comma after `user` as well.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+The standard approach requires
+adding an `<input>` tag to each `<form>` in every template.
+That's a lot of repeated code,
+even if you boil it down to a function call
+or a single partial template;
+there are likely hundreds of forms across your application.
+At the very least, that's just more boiler-plate code
+that slows development progress.
+At the worst, you may forget to add the CSRF `<input>` tag
+to some seldom-used forms and they'll efectively be broken.
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: efectively -> effectively

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+### Our approach to prevention
+We use javascript to dynamically set a cookie
+and add an `<input>` element to the form
+immediately before it's submitted.
+Because we just have to prove that we can alter or read cookies,
+we don't need to depend on the server-side for setting the cookie;
+the client can do it just fine.
+We hook into forms by replacing the `form.submit()` function
+and listening for the `onsubmit` event.
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md:70]: listening -> listen

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+This frees us from worrying about CSRF inputs and tokens in our templates.
+It reduces the boiler-plate code for creating new pages and forms
+and reduces the chance of mistakes.
+
+Here's the entire javascript implementation of our CSRF protection (Github [gist][gist]):
+{% gist 6060418 %}
+
+*Note: we use Mootools, but refactoring for no jQuery or no library would be
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: Comma splice.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+fairly straightforward.*
+
+## How we deployed this
+We designed it, tested it extensively, ran our test suite many times,
+then soft-deployed it, silently reporting CSRF failures to a database.
```
Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: I would change this to:

> and then soft-deployed it -- silently reporting CSRF failures to a database.

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 9 above)
Comment (danielbeardsley) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: So, I'd never used spell checking in vim until _just_ now... it's awesome: `:set spell spelllang=en_us` and `:help spell`

All these errors prompted me to investigate.

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 10 above)
Comment (danielbeardsley) [inline _posts/2013-07-21-csrf-how-we-do-it.md:70]: I think `We do blah by replacing X and listening for Y` makes more sense that `We do blah by replacing X and listen to Y`

Comment (brotatos) [inline _posts/2013-07-21-csrf-how-we-do-it.md:70]: Sounds good to me; the sentence structure simply needs to stay parallel.

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+title: "CSRF - How we protect ifixit.com from request forgery"
+author: Daniel Beardsley
+author_url: http://github.com/danielbeardsley
+summary: We've come up with a minimalist method of protecting our site
+         from cross site request forgery.
+---
+
+We've come up with a minimalist method of
+protecting our site from cross site request forgery ([CSRF][csrf]),
+and we'd like to share how we do it.
```
Comment (marczych) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: I don't really like the end of this sentence. Maybe splitting it into 2 sentences would be best? e.g. We've come up with a minimalist method of protecting our site from cross site request forgery ([CSRF][csrf]). Here's how we do it.

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 10 above)
Comment (marczych) [inline _posts/2013-07-21-csrf-how-we-do-it.md:70]: I also thought it was wrong the first time I read it. It is either:
`We hook into X and listen for Y.` or
`We hook into blah by replacing X and listening for Y.`

I think the most obvious interpretation as you read it for the first time is `We hook into X and listening for Y.` Could we rewrite it to remove the ambiguity?

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+*Note: we use Mootools &mdash; refactoring for no jQuery or no library would be
+fairly straightforward.*
+
+## How we deployed this
+We designed it, tested it extensively, ran our test suite many times,
+then soft-deployed it &mdash; silently reporting CSRF failures to a database.
+Once we'd worked out a few issues
+and were confident that we'd covered all our bases,
+we flipped the switch on enforcement
```
Comment (marczych) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: we flipped the **enforcement** switch on?

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+and it's been working well since.
```
Comment (marczych) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: and it's been working well **ever** since.?

File: _posts/2013-07-21-csrf-how-we-do-it.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+we don't need to depend on the server-side for setting the cookie;
+the client can do it just fine.
+We hook into forms by replacing the `form.submit()` function
+and listening for the `onsubmit` event.
+This frees us from worrying about CSRF inputs and tokens in our templates.
+It reduces the boiler-plate code for creating new pages and forms
+and reduces the chance of mistakes.
+
+Here's the entire javascript implementation of our CSRF protection (Github [gist][gist]):
+{% gist 6060418 %}
```
Comment (marczych) [inline _posts/2013-07-21-csrf-how-we-do-it.md:79]: The gist finds all forms and adds the csrf input when the page initially loads (`domready` event). Is it worth mentioning that something special needs to happen when dynamically creating forms in JS? (Something special needs to happen right?)

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 10 above)
Comment (danielbeardsley) [inline _posts/2013-07-21-csrf-how-we-do-it.md:70]: I still don't understand, I must be reading my own sentence oddly. What is ambiguous about `We do blah by replacing X and listening for Y`?. The rewrite of `We hook into X and listen for Y.` leaves out the `by replacing the ...` part which explains one of the ways we "hook" into forms.

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 15 above)
Comment (danielbeardsley) [inline _posts/2013-07-21-csrf-how-we-do-it.md]: Both are valid, I went with the terser, but it does sound like `ever` was unintentionally left out, I'll add it.

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 16 above)
Comment (danielbeardsley) [inline _posts/2013-07-21-csrf-how-we-do-it.md:79]: Good point. I'll add that.

File: _posts/2013-07-21-csrf-how-we-do-it.md (hunk 10 above)
Comment (marczych) [inline _posts/2013-07-21-csrf-how-we-do-it.md:70]: The `X` in `We hook into X and listen for Y.` is `forms by replacing the 'form.submit' function`.

I've read the sentence too many times now so I can't decide whether it's fine or not.

**Comment by danielbeardsley** [comment]
Ok... I'm gonna merge this.

**Pull request merged**